doc/modules/gaussian_process.rst

```diff
@@ -8,45 +8,514 @@ Gaussian Processes
+data to define a likelihood function. Based on Bayes theorem, a (Gaussian)
+posterior distribution over target functions is defined, whose mean is used
+for prediction.
+
+A major difference is that GPR can choose the kernel's hyperparameters based
+on gradient-ascent on the marginal likelihood function while KRR needs to
+perform a grid search on a cross-validated loss function (mean-squared error
+loss). A further difference is that GPR learns a generative, probabilistic
+model of the target function and can thus provide meaningful confidence
+intervals and posterior samples along with the predictions while KRR only
+provides predictions.
+
+The following figure illustrates both methods on an artificial dataset, which
+consists of a sinusoidal target function and strong noise. The figure compares
+the learned model of KRR and GPR based on a ExpSineSquared kernel, which is
+suited for learning periodic functions. The kernel's hyperparameters control
+the smoothness (l) and periodicity of the kernel (p). Moreover, the noise level
+of the data is learned explicitly by GPR by an additional WhiteKernel component
+in the kernel and by the regularization parameter alpha of KRR.
+
+.. figure:: ../auto_examples/gaussian_process/images/plot_compare_gpr_krr_001.png
+   :target: ../auto_examples/gaussian_process/plot_compare_gpr_krr.html
+   :align: center
+
+The figure shows that both methods learn reasonable models of the target
+function. GPR correctly identifies the periodicity of the function to be
+roughly 2*pi (6.28), while KRR chooses the doubled periodicity 4*pi. Besides
+that, GPR provides reasonable confidence bounds on the prediction which are not
+available for KRR. A major difference between the two methods is the time
+required for fitting and predicting: while fitting KRR is fast in principle,
+the grid-search for hyperparameter optimization scales exponentially with the
+number of hyperparameters ("curse of dimensionality"). The gradient-based
+optimization of the parameters in GPR does not suffer from this exponential
+scaling and is thus considerable faster on this example with 3-dimensional
+hyperparameter space. The time for predicting is similar; however, generating
+the variance of the predictive distribution of GPR takes considerable longer
+than just predicting the mean.
+
+GPR on Mauna Loa CO2 data
+-------------------------
+
+This example is based on Section 5.4.3 of "Gaussian Processes for Machine
+Learning" [RW2006]_. It illustrates an example of complex kernel engineering and
+hyperparameter optimization using gradient ascent on the
+log-marginal-likelihood. The data consists of the monthly average atmospheric
+CO2 concentrations (in parts per million by volume (ppmv)) collected at the
+Mauna Loa Observatory in Hawaii, between 1958 and 1997. The objective is to
+model the CO2 concentration as a function of the time t.
+
+The kernel is composed of several terms that are responsible for explaining
```

Comment:
This is basically "howto" on kernel engineering - fantastic! In a future PR maybe we could have more on this - or is there an existing reference for practitioners that could be shown?